### PR TITLE
Sync Dyson Sphere object

### PR DESCRIPTION
**Pull request by Baldie-dev**
This is part of the #73 synchronization task.

The first iteration that works and will be enhanced in the future.

Completed:
- every time player enters the star system, it will request Dyson Sphere from the server if it is missing on the client-side.
- server serves Dyson sphere to clients when they request it.

[Youtube video](https://www.youtube.com/watch?v=U9TwGIr34yA) (The displayed error is not related to this)

### CODE REVIEW

File: NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+            using (MemoryStream ms = new MemoryStream())
+            {
+                using (LZ4Stream ls = new LZ4Stream(ms, CompressionMode.Compress))
+                using (BufferedStream bs = new BufferedStream(ls, 8192))
+                using (BinaryWriter bw = new BinaryWriter(bs))
+                {
+                    dysonSphere.Export(bw);
+                }
+                conn.SendPacket(new DysonSphereData(packet.StarIndex, ms.ToArray()));
```
Comment (Therzok) [inline NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs:27]: Seems weird to have to convert this to an array when we just want to transmit data over a stream. Can't we use CopyTo after resetting the position to 0

File: NebulaModel/Packets/Universe/DysonSphereResponse.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+        public DysonSphereResponse() { }
+
+        public DysonSphereResponse(Dictionary<int, byte[]> dysonSphereData)
+        {
+            this.StarIndexes = dysonSphereData.Keys.ToArray();
+
+            // Can't use a jagged array because LNL serializer says no, so flattening and separate offset array it is
+            // TODO: Possibly register a type with the serializer instead of doing this manually
+            this.DysonSphereBytes = dysonSphereData.Values.SelectMany(x => x).ToArray();
```
Comment (Therzok) [inline NebulaModel/Packets/Universe/DysonSphereResponse.cs:20]: Isn't this the biggest chunk of the memory used by the save file? Allocating arrays here feels really heavy.

File: NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs (hunk 1 above)
Comment (Baldie-dev) [reply NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs:27]: Actually, I was just following the same approach that is already in [FactoryLoadRequestProcessor](https://github.com/hubastard/nebula/blob/0f1ba59d099e9d3b908fcf9c78c25ed982aed3bd/NebulaHost/PacketProcessors/Planet/FactoryLoadRequestProcessor.cs#L28) and [PlanetDataRequestProcessor](https://github.com/hubastard/nebula/blob/0f1ba59d099e9d3b908fcf9c78c25ed982aed3bd/NebulaHost/PacketProcessors/Planet/PlanetDataRequestProcessor.cs#L59).

File: NebulaModel/Packets/Universe/DysonSphereResponse.cs (hunk 2 above)
Comment (Baldie-dev) [reply NebulaModel/Packets/Universe/DysonSphereResponse.cs:20]: True, the big full Sphere can have around 9MB and a small full sphere around 5MB. I was following [PlanetDataResponse](https://github.com/hubastard/nebula/blob/0f1ba59d099e9d3b908fcf9c78c25ed982aed3bd/NebulaModel/Packets/Planet/PlanetDataResponse.cs#L21). But I can try to optimize it.

File: NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs (hunk 1 above)
Comment (Therzok) [reply NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs:27]: Understood, I can take a look at this after, no worries.

File: NebulaModel/Packets/Universe/DysonSphereResponse.cs (hunk 2 above)
Comment (Baldie-dev) [reply NebulaModel/Packets/Universe/DysonSphereResponse.cs:20]: Oh, I am not even using this class anymore, so I will check it and remove it :)

Comment (Baldie-dev) [reply NebulaModel/Packets/Universe/DysonSphereResponse.cs]: removed.

File: NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs (hunk 1 above)
Comment (hubastard) [reply NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs:27]: @Therzok Were you implying that you will look at this after the merge, or you wanted to take a look before I merge this?

File: NebulaPatcher/Patches/Dynamic/PlanetModelingManager_Patch.cs [hunk 3]
```diff
@@ -66,6 +67,12 @@ public static bool RequestLoadStar(StarData star)
 
             InternalLoadPlanetsRequestGenerator(star.planets);
 
+            // Request initial dysonSphere data
+            if (GameMain.data.dysonSpheres[star.index] == null)
+            {
+                Log.Info($"Requesting DysonSphere for system {star.displayName} (Index: {star.index})");
+                LocalPlayer.SendPacket(new DysonSphereLoadRequest(star.index));
+            }
```
Comment (hubastard) [inline NebulaPatcher/Patches/Dynamic/PlanetModelingManager_Patch.cs:75]: So, if I understand correctly, we only request it once. We will add incremental updates later ?

**Review by hubastard — APPROVED** [review]

File: NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs (hunk 1 above)
Comment (Therzok) [reply NebulaHost/PacketProcessors/Universe/DysonSphereRequestProcessor.cs:27]: After

**Pull request merged**